### PR TITLE
Multi-Lora support

### PR DESCRIPTION
**Pull request by yuslepukhin**

### CODE REVIEW

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 1]
```diff
@@ -2022,7 +2029,26 @@ including arg name, arg type (contains both type and shape).)pbdoc")
...
+                 total_entries += adapter->GetParamNum();
+               }
+               feeds.reserve(total_entries);
+
+               // Append necessary inputs for active adapters
+               for (const auto* adapter : run_options->active_adapters_) {
+                 auto [begin, end] = adapter->GetParamIterators();
+                 for (; begin != end; ++begin) {
+                   const auto& [name, param] = *begin;
+                   std::cout << name << ':';
```
Comment (yuslepukhin) [inline onnxruntime/python/onnxruntime_pybind_state.cc:2044]: Remove debug output

File: onnxruntime/test/testdata/lora/two_params_lora_model.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
+import os
+
+import numpy as np
+import onnx
+
+import onnxruntime as ort
+
+model_path = "C:/dev/ort_main/onnxruntime/test/testdata/lora/two_params_lora_model.onnx"
```
Comment (yuslepukhin) [inline onnxruntime/test/testdata/lora/two_params_lora_model.py:8]: remove full path

File: onnxruntime/lora/adapter_format/adapter_schema.fbs [hunk 3]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace onnxruntime.adapters;
+
+// Tensor
+enum TensorDataType : int32 {
```
Comment (baijumeswani) [inline onnxruntime/lora/adapter_format/adapter_schema.fbs:7]: You can avoid redefining this by using the existing definition inside ort.fbs: https://github.com/microsoft/onnxruntime/blob/d4692835bf7affb51afa83b47159f43e7526ac9d/onnxruntime/core/flatbuffers/schema/ort.fbs#L46

Example of how to reuse the existing fbs: https://github.com/microsoft/onnxruntime/blob/d4692835bf7affb51afa83b47159f43e7526ac9d/onnxruntime/core/flatbuffers/schema/ort_training_checkpoint.fbs#L4

File: onnxruntime/lora/adapter_format/adapter_schema.fbs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  FLOAT8E4M3FN = 17,
+  FLOAT8E4M3FNUZ = 18,
+  FLOAT8E5M2 = 19,
+  FLOAT8E5M2FNUZ = 20,
+}
+
+// For simplicity, we will have only have one data field
+// - raw_data for all primitive types.
+// We do not foresee strings as parameters.
+table Parameter {
```
Comment (baijumeswani) [inline onnxruntime/lora/adapter_format/adapter_schema.fbs:34]: You could consider reusing `Tensor` from ort.fbs: https://github.com/microsoft/onnxruntime/blob/d4692835bf7affb51afa83b47159f43e7526ac9d/onnxruntime/core/flatbuffers/schema/ort.fbs#L150-L166

There already are many utilities around ort tensor loading and saving which you could take advantage of.

Since this data type is opaque to the user, the user doesn't need to know what is used under the hood (whether it is a Parameter or a Tensor).

Comment (yuslepukhin) [reply onnxruntime/lora/adapter_format/adapter_schema.fbs:34]: See the comment above.

File: onnxruntime/lora/adapter_format/adapter_schema.fbs (hunk 3 above)
Comment (yuslepukhin) [reply onnxruntime/lora/adapter_format/adapter_schema.fbs:7]: The adapters format is designed to be standardized and possibly moved to ONNX.
As such it must not be a blackbox for end users but must be publicly defined.
It cannot depend on ORT private format which as subject to change release to release. Said that, the only thing which potentially can be shared in type enums.
It also does not share the namespace which was decided upon by PMs.
For the above reasons it does not and cannot share any utility code, which I checked, amounts only to a few lines of code (Save/Load string) everything else is different.

File: onnxruntime/test/testdata/lora/two_params_lora_model.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
+import os
```
Comment (baijumeswani) [inline onnxruntime/test/testdata/lora/two_params_lora_model.py:1]: nit: license

File: onnxruntime/python/onnxruntime_pybind_lora.cc [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+          },
+          R"pbdoc(The function returns an instance of the class that contains a dictionary of name -> numpy arrays)pbdoc");
+
+  py::class_<lora::LoraAdapter> lora_adapter_binding(m, "LoraAdapter");
+  lora_adapter_binding.def(py::init())
+      .def("Load", [](lora::LoraAdapter* adapter, const std::wstring& file_path) { adapter->MemoryMap(file_path); }, R"pbdoc(Memory map the specified file as LoraAdapter)pbdoc");
+}
+
+}  // namespace python
+}  // namespace onnxruntime
```
Comment (baijumeswani) [inline onnxruntime/python/onnxruntime_pybind_lora.cc:151]:
```suggestion
}  // namespace onnxruntime

```

**Review by baijumeswani — COMMENTED** [review]
Minor comments. Otherwise, the pull-request looks good.

File: include/onnxruntime/core/framework/run_options.h [hunk 7]
```diff
@@ -40,6 +48,8 @@ struct OrtRunOptions {
   // /include/onnxruntime/core/session/onnxruntime_run_options_config_keys.h
   onnxruntime::ConfigOptions config_options;
 
+  onnxruntime::InlinedVector<const onnxruntime::lora::LoraAdapter*> active_adapters_;
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/run_options.h:51]: nit: struct data member naming

https://google.github.io/styleguide/cppguide.html#Variable_Names

```suggestion
  onnxruntime::InlinedVector<const onnxruntime::lora::LoraAdapter*> active_adapters;
```

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 8]
```diff
@@ -4670,6 +4671,41 @@ struct OrtApi {
...
+  /** \brief Create an OrtLoraAdapter
+   *
+   * The function attempts to locate file specified by adapter_file_path, read it and create an OrtLoraAdapter
+   * instance. The adapter_file_path should be a valid absolute path to a file that contains a valid Lora Adapter
+   * format. The function attempts to validate the format at load time. The file will always be memory mapped, unless
+   * the platform does not support memory mapping, in which case the file will be read into memory.
+   *
+   * \param[in] adapter_file_path Absolute file path to the adapter file.
+   * \param[in] allocator optional pointer to a device allocator. If specified
+   *            data is copied to the device at some point before Run() is invoked, if nullptr data stays on CPU.
```
Comment (edgchen1) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:4684]:
```suggestion
   * \param[in] allocator optional pointer to a device allocator. If specified,
   *            data is copied to the device at some point before Run() is invoked. If nullptr, data stays on CPU.
```

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 9]
```diff
@@ -4670,6 +4671,41 @@ struct OrtApi {
...
                   size_t num_external_initializer_files);
+
+  /** \brief Create an OrtLoraAdapter
+   *
+   * The function attempts to locate file specified by adapter_file_path, read it and create an OrtLoraAdapter
+   * instance. The adapter_file_path should be a valid absolute path to a file that contains a valid Lora Adapter
+   * format. The function attempts to validate the format at load time. The file will always be memory mapped, unless
+   * the platform does not support memory mapping, in which case the file will be read into memory.
+   *
+   * \param[in] adapter_file_path Absolute file path to the adapter file.
```
Comment (edgchen1) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:4682]: what's the reason for not also allowing relative paths here? is this constraint inconsistent other path parameters, like CreateSession()'s model_path?

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 10]
```diff
@@ -4670,6 +4671,41 @@ struct OrtApi {
...
+   */
+  ORT_CLASS_RELEASE(LoraAdapter);
+
+  /** \brief Set the active Lora Adapter for the run options
+   *
+   * The function sets the active Lora Adapter for the run options. The Lora Adapter must be created with
+   * OrtApi::CreateLoraAdapter. The Lora Adapter will be used by the session to run the model.
+   * The instance of the OrtRunOptions will then can be used to customize the OrtSession::Run() calls.
+   * More than one OrtLoraAdapter can be set active at the same time. Lora Parameters that belong to difference
+   * Lora adapters that will be active at the same time must not overlap.
```
Comment (edgchen1) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:4702]:
```suggestion
   * The instance of the OrtRunOptions can then be used to customize the Run() calls.
   * More than one OrtLoraAdapter can be active at the same time. Lora Parameters that belong to different
   * Lora adapters that will be active at the same time must not overlap.
```

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 11]
```diff
@@ -4670,6 +4671,41 @@ struct OrtApi {
...
+   * This setting does not affect RunWithBinding.
+   *
+   * \param[in] options OrtRunOptions instance
+   * \param[in] adapter OrtLoraAdapter instance
+   */
+  ORT_API2_STATUS(RunOptionsSetActiveLoraAdapter, _Inout_ OrtRunOptions* options, _In_ const OrtLoraAdapter* adapter);
```
Comment (edgchen1) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:4708]: so to get multiple active adapters, this function should be called multiple times?

File: onnxruntime/core/framework/run_options.cc [hunk 12]
```diff
@@ -60,3 +62,12 @@ ORT_API_STATUS_IMPL(OrtApis::AddRunConfigEntry, _Inout_ OrtRunOptions* options,
                     _In_z_ const char* config_key, _In_z_ const char* config_value) {
   return onnxruntime::ToOrtStatus(options->config_options.AddConfigEntry(config_key, config_value));
 }
+
+ORT_API_STATUS_IMPL(OrtApis::RunOptionsSetActiveLoraAdapter, _Inout_ OrtRunOptions* options,
+                    const _In_ OrtLoraAdapter* adapter) {
+  API_IMPL_BEGIN
+  auto* lora_adapter = reinterpret_cast<const onnxruntime::lora::LoraAdapter*>(adapter);
+  options->active_adapters_.push_back(lora_adapter);
```
Comment (edgchen1) [inline onnxruntime/core/framework/run_options.cc:70]: regarding naming, would "AddActiveLoraAdapter" be more appropriate than "SetActiveLoraAdapter"?

**Review by edgchen1 — COMMENTED** [review]
started reviewing, will continue later

File: onnxruntime/python/onnxruntime_pybind_lora.cc (hunk 6 above)
Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_lora.cc:151]: lintrunner would have its own way.

File: include/onnxruntime/core/framework/run_options.h [hunk 13]
```diff
@@ -5,9 +5,17 @@
...
+
+#include "core/common/inlined_containers_fwd.h"
 #include "core/session/onnxruntime_c_api.h"
 #include "core/framework/config_options.h"
 
+namespace onnxruntime {
+namespace lora {
+class LoraAdapter;
+}
+}  // namespace onnxruntime
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply include/onnxruntime/core/framework/run_options.h:17]: I am not against it, it just we never do it in this codebase.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 14]
```diff
@@ -4670,6 +4671,41 @@ struct OrtApi {
...
+   * the platform does not support memory mapping, in which case the file will be read into memory.
+   *
+   * \param[in] adapter_file_path Absolute file path to the adapter file.
+   * \param[in] allocator optional pointer to a device allocator. If specified
+   *            data is copied to the device at some point before Run() is invoked, if nullptr data stays on CPU.
+   *            The data would still be copied to device if required by the model at inference time.
+   * \param[out] out A pointer to a newly created OrtLoraAdapter instance. Must be released with
+   *                  OrtApi::ReleaseLoraAdapter.
+   */
+  ORT_API2_STATUS(CreateLoraAdapter, const ORTCHAR_T* adapter_file_path, _In_ OrtAllocator* allocator,
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:4689]: See internal design doc comments.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 9 above)
Comment (yuslepukhin) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:4682]: It was a security consideration. It also a part of the design.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 11 above)
Comment (yuslepukhin) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:4708]: Yes.

File: cmake/onnxruntime_session.cmake [hunk 15]
```diff
@@ -30,7 +30,8 @@ endif()
 source_group(TREE ${REPO_ROOT} FILES ${onnxruntime_session_srcs})
 
 onnxruntime_add_static_library(onnxruntime_session ${onnxruntime_session_srcs})
-onnxruntime_add_include_to_target(onnxruntime_session onnxruntime_common onnxruntime_framework onnx onnx_proto ${PROTOBUF_LIB} flatbuffers::flatbuffers Boost::mp11 safeint_interface nlohmann_json::nlohmann_json)
+onnxruntime_add_include_to_target(onnxruntime_session onnxruntime_common onnxruntime_framework onnxruntime_lora onnx onnx_proto ${PROTOBUF_LIB} flatbuffers::flatbuffers Boost::mp11 safeint_interface nlohmann_json::nlohmann_json)
+target_link_libraries(onnxruntime_session PRIVATE onnxruntime_lora)
```
[the remark this answers is not in the document]
Comment (snnn) [reply cmake/onnxruntime_session.cmake:34]: @yuslepukhin , if you remove this line, what would happen? onnxruntime_session  is a static library which has no link stage.  I wonder why this line is needed.

[the remark this answers is not in the document]
Comment (yuslepukhin) [reply cmake/onnxruntime_session.cmake:34]: I will try

[the remark this answers is not in the document]
Comment (yuslepukhin) [reply cmake/onnxruntime_session.cmake]: 2024-09-25T21:21:30.4712275Z /opt/rh/gcc-toolset-12/root/usr/libexec/gcc/x86_64-redhat-linux/12/ld: libonnxruntime_session.a(lora_adapters.cc.o): in function `onnxruntime::lora::LoraAdapter::InitializeParamsValues()':
2024-09-25T21:21:30.4719537Z lora_adapters.cc:(.text._ZN11onnxruntime4lora11LoraAdapter22InitializeParamsValuesEv+0x4f1): undefined reference to `onnxruntime::adapters::utils::CreateOrtValueOverLoraParameter[abi:cxx11](onnxruntime::adapters::Parameter const&)'
2024-09-25T21:21:30.4727607Z /opt/rh/gcc-toolset-12/root/usr/libexec/gcc/x86_64-redhat-linux/12/ld: libonnxruntime_session.a(lora_adapters.cc.o): in function `onnxruntime::lora::LoraAdapter::Load(std::vector<unsigned char, std::allocator<unsigned char> >)':

[the remark this answers is not in the document]
Comment (snnn) [reply cmake/onnxruntime_session.cmake]: Please open [cmake/onnxruntime_training.cmake](https://github.com/microsoft/onnxruntime/blob/50d9612bc035d9904c739cdcdedbec8394d338b8/cmake/onnxruntime_training.cmake#L110) and add onnxruntime_lora to line 142

[the remark this answers is not in the document]
Comment (snnn) [reply cmake/onnxruntime_session.cmake:34]: I am trying to find out where the cmake variable `onnxruntime_libs` was defined.  We need to add onnxruntime_lora to the list.

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (baijumeswani) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:4689]: My comment was mostly about renaming it to `CreateAdapter` instead of `CreateLoraAdapter` since this API can also be used to load other (non LoRA) adapters such as those mentioned in my comment above.

But based on the offline sync, it seems like the consensus is to use `LoraAdapter` for all types of adapters.

File: onnxruntime/core/session/onnxruntime_c_api.cc [hunk 16]
```diff
@@ -2786,7 +2832,7 @@ ORT_API(const char*, OrtApis::GetVersionString) {
   return ORT_VERSION;
 }
 
-ORT_API(const char*, OrtApis::GetBuildInfoString) {
+const char* ORT_API_CALL OrtApis::GetBuildInfoString() noexcept {
```
Comment (edgchen1) [inline onnxruntime/core/session/onnxruntime_c_api.cc:2835]: why this change? doesn't ORT_API already include `ORT_API_CALL` and `NO_EXCEPTION`?

https://github.com/microsoft/onnxruntime/blob/7880342e5e50159f59b9f9c0bd0a74ecc1542549/include/onnxruntime/core/session/onnxruntime_c_api.h#L143

File: onnxruntime/core/session/lora_adapters.cc [hunk 17]
```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "lora_adapters.h"
```
Comment (edgchen1) [inline onnxruntime/core/session/lora_adapters.cc:4]: general: use include path relative to project source directory

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: onnxruntime/lora/adapter_format_version.h [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+#include <algorithm>
+#include <array>
+
+namespace onnxruntime {
+namespace adapters {
+
+// The current model versions for saving lora parameters in flatbuffers
+// Once this version is updated, the kSupportedAdapterFormatVersions in IsGenAiLoraFormatModelBytes
+// below will also need to be updated.
+// See src/flatbuffers/schema/README.md for more details on versioning.
```
Comment (edgchen1) [inline onnxruntime/lora/adapter_format_version.h:15]: double check comment. e.g., readme path, "IsGenAiLoraFormatModelBytes"

File: cmake/onnxruntime_session.cmake (hunk 15 above)
[the remark this answers is not in the document]
Comment (snnn) [reply cmake/onnxruntime_session.cmake:34]: There is no such variable we can use. You may need to manually add onnxruntime_lora to each target that has link errors.

**Review by baijumeswani — APPROVED** [review]

File: onnxruntime/python/onnxruntime_pybind_ortvalue.cc [hunk 19]
```diff
@@ -142,6 +142,28 @@ void addOrtValueMethods(pybind11::module& m) {
...
+        const auto element_size = element_type->Size();
+        if (narrow<size_t>(data.itemsize()) != element_size) {
+          ORT_THROW("Items size in the incoming aray: ", data.itemsize(),
+                    " specified by onnxtype: ", element_size);
+        }
+
+        auto cpu_allocator = GetAllocator();
+        auto ort_value = std::make_unique<OrtValue>();
+        Tensor::InitOrtValue(element_type, GetShape(data),
+                             const_cast<void*>(data.data()), cpu_allocator->Info(), *ort_value);
```
Comment (edgchen1) [inline onnxruntime/python/onnxruntime_pybind_ortvalue.cc:164]: is const_cast needed? `data` is not const, but I failed to find the signature for py::array::data in a brief search for documentation.

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 20]
```diff
@@ -1901,7 +1903,12 @@ RunOptions instance. The individual calls will exit gracefully and return an err
 
             return value;
           },
-          R"pbdoc(Get a single run configuration value using the given configuration key.)pbdoc");
+          R"pbdoc(Get a single run configuration value using the given configuration key.)pbdoc")
+      .def(
+          "set_adapter_active", [](RunOptions* options, lora::LoraAdapter* adapter) {
+            options->active_adapters.push_back(adapter);
+          },
+          R"pbdoc(Activates the specified lora adapter)pbdoc");
```
Comment (edgchen1) [inline onnxruntime/python/onnxruntime_pybind_state.cc:1911]: naming suggestion on adding vs setting
```suggestion
          "add_active_adapter", [](RunOptions* options, lora::LoraAdapter* adapter) {
            options->active_adapters.push_back(adapter);
          },
          R"pbdoc(Adds the specified lora adapter as an active adapter)pbdoc");
```

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 21]
```diff
@@ -2022,7 +2029,25 @@ including arg name, arg type (contains both type and shape).)pbdoc")
               const std::map<std::string, const py::object>& pyfeeds, RunOptions* run_options = nullptr)
                -> py::list {
              NameMLValMap feeds;
-             feeds.reserve(pyfeeds.size());
+             if (run_options != nullptr && !run_options->active_adapters.empty()) {
```
Comment (edgchen1) [inline onnxruntime/python/onnxruntime_pybind_state.cc:2032]: maybe also add a warning about ignoring any active adapters in run options for the other run methods.

File: onnxruntime/test/lora/lora_test.cc [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+  }
+};
+
+}  // namespace
+
+TEST(LoraAdapterTest, Load) {
+  // Test different data types
+  const auto data_types = gsl::make_span(adapters::EnumValuesTensorDataType());
+  for (size_t i = 1, size = data_types.size(); i < size; ++i) {
+    if (i == 8 || i == 9 || i == 14 || i == 15 || (i > 16 && i < 21))
```
Comment (edgchen1) [inline onnxruntime/test/lora/lora_test.cc:183]: could we cast `i` to the enum type and compare with enum values?

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 23]
```diff
@@ -4402,6 +4402,120 @@ TEST(CApiTest, RunAsyncFail) {
...
+
+  std::vector<uint8_t> buffer(adapter_size);
+  adapter_file.seekg(0, std::ios::beg);
+  adapter_file.read(reinterpret_cast<char*>(buffer.data()), adapter_size);
+  adapter_file.close();
+
+  return Ort::LoraAdapter::CreateLoraAdapterFromArray(buffer.data(), buffer.size(), nullptr);
+}
+
+TEST(CApi, RunWithLoraAdapterFromFile) {
```
Comment (edgchen1) [inline onnxruntime/test/shared_lib/test_inference.cc:4470]:
```suggestion
TEST(CApiTest, RunWithLoraAdapterFromFile) {
```

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 24]
```diff
@@ -4402,6 +4402,120 @@ TEST(CApiTest, RunAsyncFail) {
...
+  auto adapter = CreateAdapterFromFile();
+  TestRunWithLoraAdapter(adapter);
+}
+
+TEST(CApi, RunWithLoraAdapterFromArray) {
```
Comment (edgchen1) [inline onnxruntime/test/shared_lib/test_inference.cc:4475]:
```suggestion
TEST(CApiTest, RunWithLoraAdapterFromArray) {
```

File: onnxruntime/python/onnxruntime_inference_collection.py [hunk 25]
```diff
@@ -32,6 +32,52 @@ def get_ort_device_type(device_type: str, device_index) -> C.OrtDevice:
         raise Exception("Unsupported device type: " + device_type)
 
 
+class AdapterFormat:
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_inference_collection.py:35]: It would be harder to keep up to date, because it would not be used.

File: onnxruntime/python/onnxruntime_pybind_ortvalue.cc (hunk 19 above)
Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_ortvalue.cc:164]: Yes.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 26]
```diff
@@ -4670,6 +4671,57 @@ struct OrtApi {
                   _In_reads_(num_external_initializer_files) char* const* external_initializer_file_buffer_array,
                   _In_reads_(num_external_initializer_files) const size_t* external_initializer_file_lengths,
                   size_t num_external_initializer_files);
+
+  /** \brief Create an OrtLoraAdapter
+   *
+   * The function attempts to locate file specified by adapter_file_path, read it and create an OrtLoraAdapter
+   * instance. The adapter_file_path should be a valid absolute path to a file that contains a valid Lora Adapter
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:4678]: I reverted to just path, now allow relative paths like in Ort::Session

**Review by baijumeswani — APPROVED** [review]

File: onnxruntime/lora/adapter_format_utils.cc [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+  const auto data_type = param.data_type();
+  gsl::span<const int64_t> shape_span(param.dims()->data(), param.dims()->size());
+
+  static const OrtMemoryInfo cpu_meminfo(CPU, OrtAllocatorType::OrtDeviceAllocator);
+
+  auto elem_type = DataTypeImpl::TensorTypeFromONNXEnum(static_cast<int32_t>(data_type))->GetElementType();
+  // const_cast is necessery due to Tensor class API
+  Tensor::InitOrtValue(elem_type,
+                       TensorShape(shape_span),
+                       const_cast<uint8_t*>(param.raw_data()->data()),
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/lora/adapter_format_utils.cc:104]: Discussed offline. OrtValue can not be const and it has no bearing on the non-const data passed. It is dictated by Tensor constructor API which does not take const data.

File: onnxruntime/python/onnxruntime_inference_collection.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/python/onnxruntime_inference_collection.py:35]: sorry, I could have been more explicit. could we also use the generated python bindings instead of implementing our own python adapter format related APIs?

File: onnxruntime/python/onnxruntime_pybind_state.cc (hunk 21 above)
Comment (edgchen1) [reply onnxruntime/python/onnxruntime_pybind_state.cc:2046]: also `run_with_iobinding`

**Review by edgchen1 — DISMISSED** [review]

**Comment by ranjitshs** [comment]
@snnn @yuslepukhin 
I see some tests failures in AIX. I will debug and raise PR if I find any fix.

```
4: [ RUN      ] CApiTest.RunWithLoraAdapterFromFile
4: unknown file: Failure
4: C++ exception with description "/home/buildusr/jenkins/workspace/onnxruntime-openxl/onnxruntime/onnxruntime/core/common/safeint.h:17 static void SafeIntExceptionHandler<onnxruntime::OnnxRuntimeException>::SafeIntOnOverflow() Integer overflow
4: " thrown in the test body.
4: 
4: [  FAILED  ] CApiTest.RunWithLoraAdapterFromFile (0 ms)
4: [ RUN      ] CApiTest.RunWithLoraAdapterFromArray
4: unknown file: Failure
4: C++ exception with description "/home/buildusr/jenkins/workspace/onnxruntime-openxl/onnxruntime/onnxruntime/core/common/safeint.h:17 static void SafeIntExceptionHandler<onnxruntime::OnnxRuntimeException>::SafeIntOnOverflow() Integer overflow
4: " thrown in the test body.
4: 
4: [  FAILED  ] CApiTest.RunWithLoraAdapterFromArray (0 ms)
4: [ RUN      ] CApiTest.RunBaseLoraModel
```

**Comment by snnn** [comment]
@yuslepukhin , and there is a compile error on Android ARM32:
```
/onnxruntime_src/onnxruntime/test/shared_lib/test_inference.cc:4460:31: error: implicit conversion loses integer precision: 'streamoff' (aka 'long long') to 'const size_t' (aka 'const unsigned int') [-Werror,-Wshorten-64-to-32]
 4460 |   const size_t adapter_size = adapter_file.tellg();
```
